### ISSUE TITLE
Secondary indexes should be used for queries with FINAL

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#70292]
**Describe the situation**

Filtering by secondary indexes is disabled for queries with FINAL, because skipping a granule can change the presence or values of records in other granules for the same sorting key as for skipped records.

See https://github.com/ClickHouse/ClickHouse/pull/34243

**How to solve**

If a query contains FINAL, secondary indices should retain granules that could contain records with the same value of the sorting key as in the records in granules that are not going to be skipped.

As a result, they will skip less amount of granules because skipped ranges will be smaller.

**Example**

Let's suppose we have a table
```
CREATE TABLE table
(
id String, 
payload String, 
PRIMARY KEY (id), 
INDEX ix (payload) TYPE = set(100)
)
ENGINE = ReplacingMergeTree
```
containing the following granules:
```
granule 1:    primary key: A    ix: (abc, hello)
id  payload
---
A abc
A hello
B abc

granule 2:    primary key: B    ix: (world, xyz)
---
B world
B xyz

granule 3:    primary key: B    ix: (hello)
---
B hello

granule 4:    primary key: C    ix: (abc)
---
C abc

granule 5:    primary key: C    ix: (def)
---
C def

granule 6:    primary key: D    ix: (world)
---
D world
```

If we do `SELECT * FROM table`, it returns:
```
A abc
A hello
B abc
B world
B xyz
B hello
C abc
C def
D world
```

If we do `SELECT * FROM table FINAL`, it returns:
```
A hello
B hello
C def
D world
```

If we do `SELECT * FROM table WHERE payload = 'world'`, it returns:
```
B world
D world
```

If we do `SELECT * FROM table FINAL WHERE payload = 'world'` with the setting `use_skip_indexes_if_final`, it returns:
```
B world
D world
```
while it should return
```
D world
```
because the index skipped the granule 3 which contained a replacement for key B, while it should not.

Let's look at the granules, under the filter `payload = 'world'`:
```
SKIPPED granule 1:    primary key: A    possible values of PK: [A..B]    ix: (abc, hello)
PASSED  granule 2:    primary key: B    possible values of PK: [B..B]    ix: (world, xyz)
SKIPPED granule 3:    primary key: B    possible values of PK: [B..C]    ix: (hello)
SKIPPED granule 4:    primary key: C    possible values of PK: [C..C]    ix: (abc)
SKIPPED granule 5:    primary key: D    possible values of PK: [C..D]    ix: (def)
PASSED  granule 6:    primary key: D    possible values of PK: [D..?]    ix: (world)
```

Granule 1 is skipped. But it could contain the same values of PK (B) as in the passed granule 2 - then it should not be skipped.
Granule 3 is skipped. But it could contain the same values of PK (B) as in the passed granule 2 - then it should not be skipped.
Granule 4 is skipped. It is ok because the values of PK in this granule (C) don't intersect with any of the passed granules.
Granule 5 is skipped. But it could contain the same values of PK (D) as in the passed granule 6 - then it should not be skipped.

The correct application of the index should look like this:
```
PASSED  granule 1:    primary key: A    possible values of PK: [A..B]    ix: (abc, hello)
PASSED  granule 2:    primary key: B    possible values of PK: [B..B]    ix: (world, xyz)
PASSED  granule 3:    primary key: B    possible values of PK: [B..C]    ix: (hello)
SKIPPED granule 4:    primary key: C    possible values of PK: [C..C]    ix: (abc)
PASSED  granule 5:    primary key: D    possible values of PK: [C..D]    ix: (def)
PASSED  granule 6:    primary key: D    possible values of PK: [D..?]    ix: (world)
```

**Comment by qoega** [comment]
I do not think it will work as you have multiple parts with intersecting primary key. The only option is to merge secondary indexes between parts to never skip primary index range that was not skipped in any of the parts.

**Comment by nickitat** [comment]
> The only option is to merge secondary indexes between parts

not necessarily the skipping indexes itself. we might implement it as a two-step process.
1. do normal scan for secondary index and for each granule collect
	* a bit of information, was it discarded by the index or not
	* if not - pk range corresponding to it into some data structure D
2. scan all granules once again and discard those that have bit=0 and their pk range doesn't intersect with the ranges accumulated in D

i see D as a segment tree that we could maintain in memory and keep its size bounded by merging some close ranges into a single one when running out of memory.

---

we now able to skip final merging for ranges with non-intersecting PK and for them we don't need this special procedure at all. tbh i wouldn't be surprised if this optimisation itself will do notorious 80% of the job.

**Comment by alexey-milovidov** [comment]
Also see this: https://github.com/ClickHouse/ClickHouse/pull/70210

**Comment by filimonov** [comment]
Also similar: https://github.com/ClickHouse/ClickHouse/issues/31411

**Comment by shiyer7474** [comment]
Adding some data to the "2-pass approach" of PR #70210 and help in critical review : https://github.com/ClickHouse/ClickHouse/issues/31411#issuecomment-2370795258

**Comment by shiyer7474** [comment]
> > The only option is to merge secondary indexes between parts
> 
> not necessarily the skipping indexes itself. we might implement it as a two-step process.
> 
> 1. do normal scan for secondary index and for each granule collect
>    
>    * a bit of information, was it discarded by the index or not
>    * if not - pk range corresponding to it into some data structure D
> 2. scan all granules once again and discard those that have bit=0 and their pk range doesn't intersect with the ranges accumulated in D
> 
Similar 2 step approach is implemented in PR #70210, please do a quick check and let know your opinion.

> i see D as a segment tree that we could maintain in memory and keep its size bounded by merging some close ranges into a single one when running out of memory.
> 
Merging or collapsing of adjacent/consecutive/'near' ranges is already done when the secondary index is looked up  :-
https://github.com/ClickHouse/ClickHouse/blob/733e61cfab8e7a99d0ea5b685218a9db98de3568/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L1483
 
> we now able to skip final merging for ranges with non-intersecting PK and for them we don't need this special procedure at all. tbh i wouldn't be surprised if this optimisation itself will do notorious 80% of the job. at least in `order by ts` cases.

Can you please elaborate on which _special procedure_ and _this optimisation_? Are you referring to non-mutating secondary key columns or referring to `do_not_merge_across_partitions_select_final`?
